x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/EsqlArithmeticOperation.java

```diff
@@ -109,6 +112,49 @@ public DataType dataType() {
         return dataType;
     }
 
+    @Override
+    protected TypeResolution resolveType() {
+        TypeResolution typeResolution = super.resolveType();
+        if (typeResolution.unresolved()) {
+            return typeResolution;
+        }
+
+        return checkCompatibility();
+    }
+
+    /**
+     * Check if the two input types are compatible for this operation
+     *
+     * @return TypeResolution.TYPE_RESOLVED iff the types are compatible.  Otherwise, an appropriate type resolution error.
+     */
+    protected TypeResolution checkCompatibility() {
+        // This checks that unsigned longs should only be compatible with other unsigned longs
+        DataType leftType = left().dataType();
+        DataType rightType = right().dataType();
+        if ((rightType == DataTypes.UNSIGNED_LONG && (false == (leftType == UNSIGNED_LONG || leftType == DataTypes.NULL)))
+            || (leftType == DataTypes.UNSIGNED_LONG && (false == (rightType == UNSIGNED_LONG || rightType == DataTypes.NULL)))) {
+            return new TypeResolution(formatIncombatibleTypesMessage(symbol(), leftType.typeName(), rightType.typeName()));
+        }
+        // If the LHS is numeric, the RHS should be numeric or null
+        if (leftType.isNumeric()) {
```

Comment:
I'm wondering if we need this check at all...
Before we reach this point, `resolveType()` already checked that both arguments are numbers or dates, and for `DateTimeArithmeticOperation` we have an override that checks that they are compatible.
Maybe I'm missing some execution paths, but in that case we should probably have a symmetric check here, ie. `else if(rightType.isNumeric()) { //check left type`

Reply:
Yeah, I think you're right, this is redundant.  I'll remove it.  Good catch, thanks.